### PR TITLE
Adding Interpolation Equirectangular DA model

### PR DESCRIPTION
**Pull request by NickGeneva**
# Earth2Studio Pull Request

### Added

- **AssimilationModel Protocol** (`earth2studio/models/da/base.py`)
  - Protocol for data assimilation models with variadic DataFrame/DataArray inputs
  - Stateless `__call__` method and stateful `create_generator` interface
  - Support for pandas and cudf DataFrames

- **InterpEquirectangular Model** (`earth2studio/models/da/interp.py`)
  - Largely meant to be a toy example
  - Interpolation assimilation model for sparse observations to regular lat-lon grids
  - Interpolation methods: nearest, linear, cubic
  - Longitude periodicity handling (0-360 wrap)
  - Time tolerance filtering for observations
  - GPU support with cupy arrays returns

- **Data Assimilation Utilities** (`earth2studio/models/da/utils.py`)
  - `validate_observation_fields`: Validates required DataFrame columns
  - `filter_time_range`: Filters observations by time tolerance window with datetime64[ns] conversion
  - `dfseries_to_torch`: Converts DataFrame series to torch tensors with zero-copy dlpack transfer for cudf

### Technical Details

- Zero-copy data transfer from cudf to PyTorch via dlpack
- Automatic longitude normalization (-180,180 → 0,360) for periodic boundaries
- Device-aware data handling (CPU: numpy, GPU: cupy/cudf when available)
- Time-based observation filtering with configurable tolerance windows

### Major Points

- For dataframes / obs data structures... use pandas dataframes for cpu, cudf for gpu
- For dense ND data, use xarray numpy for cpu, cupy for gpu... DA models are trialing cupy xarray
- Introduction of the data assimilation model protocol that aims to address the flexibility issues with other models... namely allowing both dataframes and dataarrays but also multiple parameters as well as the operational patterns needed.

Btw heres a sample script for running the model in stateful mode:

```python
from datetime import datetime, timedelta

import cartopy.crs as ccrs
import matplotlib.pyplot as plt
import numpy as np
import torch
import xarray as xr
from earth2studio.data import GFS, UFSObsConv, fetch_dataframe
from earth2studio.models.da import InterpEquirectangular

# Set up device and model
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model = InterpEquirectangular(tolerance=np.timedelta64(30, "m")).to(device)
ds = UFSObsConv(tolerance=timedelta(minutes=30))
gfs_ds = GFS()

# Initial time for prediction
start_time = datetime(2024, 1, 1, 0, 0)
model.VARIABLES = ["t2m", "u10m"]

# Set up generator for 4-hour prediction
generator = model.create_generator()
generator.send(None)  # Prime the generator

results = []
# "Stateful" DA run over a time series
for hour in range(5):
    current_time = start_time + timedelta(hours=6*hour)
    time_array = np.array([current_time], dtype=np.datetime64)
    # Fetch observations for this time
    df = fetch_dataframe(
        ds,
        time_array,
        variable=list(model.input_coords()[0]['variable']),
        fields=list(model.input_coords()[0].keys()),
        device=device
    )
    # Send to generator and get interpolated result
    da = generator.send(df)
    # Fetch GFS ground truth for comparison
    gfs_da = gfs_ds(current_time, list(model.input_coords()[0]['variable']))
    results.append((current_time, da, df, gfs_da))
    
    print(f"Interpolated data shape: {da.shape}")
    print(f"Variables: {da.coords['variable'].values}")
    print(f"Time: {da.coords['time'].values}")
    print("-" * 50)

# Close generator
generator.close()
```

<details>

```python


# Plot results for each time
# Layout: n_times rows × (n_vars * 2) columns (interpolated | GFS)
n_times = len(results)
n_vars = len(model.input_coords()[0]['variable'])
n_cols = n_vars * 2  # Two columns per variable: interpolated and GFS

fig, axes = plt.subplots(
    n_times, n_cols, figsize=(28, 4 * n_times), subplot_kw={"projection": ccrs.PlateCarree()}
)

if n_times == 1:
    axes = axes.reshape(1, -1)
if n_cols == 1:
    axes = axes.reshape(-1, 1)

def extract_plot_data(data_array, var):
    """Extract 2D plot data for a variable."""
    var_data = data_array.sel(variable=var)
    
    # Get the data (handle both numpy and cupy arrays)
    if hasattr(var_data.data, "get"):
        data = var_data.data.get()
    else:
        data = var_data.data
    
    # Handle different DataArray shapes
    # GFS shape: [time, lead_time, variable, lat, lon] -> [time, variable, lat, lon] after sel
    # Interpolated shape: [time, variable, lat, lon]
    if len(data.shape) == 4:
        # GFS with lead_time dimension
        plot_data = data[0, 0, :, :]  # [time=0, lead_time=0, lat, lon]
    else:
        # Interpolated without lead_time
        plot_data = data[0, :, :]  # [time=0, lat, lon]
    
    return plot_data, var_data.coords["lon"].values, var_data.coords["lat"].values

def plot_field(ax, plot_data, lon, lat, var, time, df, title_suffix, vmin=None, vmax=None):
    """Helper function to plot a field with observations."""
    # Plot with cartopy projection
    im = ax.pcolormesh(
        lon,
        lat,
        plot_data,
        cmap="viridis",
        transform=ccrs.PlateCarree(),
        shading="nearest",
        vmin=vmin,
        vmax=vmax
    )
    
    # Filter DataFrame to observations for this variable and plot
    if df is not None:
        var_df = df[df["variable"] == var].copy()
        if len(var_df) > 0:
            # Handle both pandas and cudf DataFrames
            lon_vals = var_df["lon"].values
            lat_vals = var_df["lat"].values
            if hasattr(lon_vals, "get"):
                lon_vals = lon_vals.get()
                lat_vals = lat_vals.get()
            
            ax.scatter(
                lon_vals,
                lat_vals,
                c="red",
                s=10,
                marker="x",
                transform=ccrs.PlateCarree(),
                zorder=10,
                linewidths=1.5
            )
    
    # Add coastlines
    ax.coastlines()
    ax.gridlines()
    
    # Set extent to CONUS region
    ax.set_extent([235, 295, 25, 50], crs=ccrs.PlateCarree())
    
    ax.set_xlabel("Longitude")
    ax.set_ylabel("Latitude")
    ax.set_title(f"{var} {title_suffix} at {time.strftime('%Y-%m-%d %H:%M')}")
    plt.colorbar(im, ax=ax, label=var)
    
    return im

for time_idx, (time, da, df, gfs_da) in enumerate(results):
    for var_idx, var in enumerate(model.input_coords()[0]['variable']):
        # Extract data for both interpolated and GFS
        plot_data_interp, lon_interp, lat_interp = extract_plot_data(da, var)
        plot_data_gfs, lon_gfs, lat_gfs = extract_plot_data(gfs_da, var)
        
        # Compute common colormap range
        vmin_common = min(np.nanmin(plot_data_interp), np.nanmin(plot_data_gfs))
        vmax_common = max(np.nanmax(plot_data_interp), np.nanmax(plot_data_gfs))
        
        # Plot interpolated observations
        ax_interp = axes[time_idx, var_idx * 2]
        plot_field(ax_interp, plot_data_interp, lon_interp, lat_interp, var, time, df, "Interpolated", vmin_common, vmax_common)
        
        # Plot GFS ground truth
        ax_gfs = axes[time_idx, var_idx * 2 + 1]
        plot_field(ax_gfs, plot_data_gfs, lon_gfs, lat_gfs, var, time, None, "GFS", vmin_common, vmax_common)

plt.tight_layout()
plt.savefig("ufs_prediction_results.png", dpi=150, bbox_inches="tight")
print("Results saved to ufs_prediction_results.png")

```
</details>

ufs_prediction_results

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/earth2studio/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/earth2studio/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/earth2studio/issues) is linked to this pull request.
- [ ] Assess and address Greptile feedback (AI code review bot for guidance; use discretion, addressing all feedback is not required).

## Dependencies

- cudf (Apache)
- cupy (MIT)

### CODE REVIEW

**Comment by NickGeneva** [comment]
@greptile-ai

File: earth2studio/models/da/interp.py [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+            lat if lat is not None else np.linspace(25.0, 50.0, 101, dtype=np.float32)
+        )
+        self._lon = (
+            lon if lon is not None else np.linspace(235.0, 295.0, 241, dtype=np.float32)
+        )
+        self.interp_method = interp_method
+        self._tolerance = normalize_time_tolerance(tolerance)
+        self.register_buffer("device_buffer", torch.empty(0), persistent=False)
+
+    def input_coords(self) -> tuple[FrameSchema | CoordSystem]:
```
[the remark this answers is not in the document]
Comment (NickGeneva) [reply earth2studio/models/da/interp.py]: No, its fine as is just returning a tuple

File: earth2studio/models/da/interp.py [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+                }
+            ),
+        )
+
+    def output_coords(
+        self,
+        input_coords: tuple[CoordSystem],
+        request_time: np.ndarray,
+        **kwargs: Any,
+    ) -> CoordSystem:
```
[the remark this answers is not in the document]
Comment (NickGeneva) [reply earth2studio/models/da/interp.py]: Updated.

**Comment by NickGeneva** [comment]
@greptile-ai

File: docs/modules/models.rst [hunk 3]
```diff
@@ -96,6 +96,28 @@ etc.
    WindgustAFNO
    Identity
 
+:mod:`earth2studio.models.dx`: Data Assimilation
```
Comment (pzharrington) [inline docs/modules/models.rst]: `earth2studio.models.da`

File: docs/modules/models.rst [hunk 4]
```diff
@@ -96,6 +96,28 @@ etc.
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Diagnostic models are a class of models that do not perform time-integration.
+These may be used to map between weather/climate variables to other quantities of
+interest, used to enbable additional analysis, improve prediction accuracy, downscale,
+etc.
```
Comment (pzharrington) [inline docs/modules/models.rst]: Update description to DA -- also note that time integration/timestepping is supported in these modelsas we discussed

File: earth2studio/models/da/base.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+            can be a DataFrame (PyArrow Table or cudf DataFrame) or xarray DataArray.
+            Supports any number of arguments.
+
+        Receives
+        --------
+        tuple[pd.DataFrame | xr.DataArray, ...]
+            Observations sent via generator.send() as multiple arguments. Each
+            argument can be a DataFrame (PyArrow Table or cudf DataFrame) or xarray
+            DataArray. None is sent initially to start the generator. Supports any
+            number of arguments.
```
Comment (pzharrington) [inline earth2studio/models/da/base.py:89]: Suggest adding an example to the docstring with the `generator.send(...)` and `generator.close()` syntax

File: docs/modules/models.rst (hunk 4 above)
Comment (NickGeneva) [reply docs/modules/models.rst]: Good spot, fixed

File: docs/modules/models.rst (hunk 3 above)
Comment (NickGeneva) [reply docs/modules/models.rst]: Fixed, thanks!

File: earth2studio/models/da/base.py (hunk 5 above)
Comment (NickGeneva) [reply earth2studio/models/da/base.py:89]: Added! Thanks.
